network/src/protocols/wire/proxy_protocol.rs

```diff
@@ -0,0 +1,369 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! # An implementation of ProxyProtocol for HAProxy
+//! https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt
+//!
+//! ## Limitations
+//! - Only supports ProxyProtocol V2
+//! - Rejects ProxyProtocol V1
+//! - Only supports IPv4 & IPv6 TCP mode
+//! - All other valid connections, will just drop the address information and use the proxy's one
+//!
+//! ## Interpetations not in the spec
+//! - An address space that doesn't match the size expected is rejected e.g. too big for IPv4
+//! - Address space that's larger than the current supported requests is rejected
+
+use crate::transport::TSocket;
+use byteorder::{BigEndian, ReadBytesExt};
+use futures_util::io::AsyncReadExt;
+use libra_network_address::NetworkAddress;
+use std::{
+    io,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+};
+
+/// Unique signature required at the beginning of every PPv2 connection
+const PPV2_SIGNATURE: [u8; 12] = [
+    0x0D, 0x0A, 0x0D, 0x0A, 0x00, 0x0D, 0x0A, 0x51, 0x55, 0x49, 0x54, 0x0A,
+];
+
+// Commands
+/// Command for healthchecks and local traffic
+const PPV2_LOCAL: u8 = 0x20;
+/// Command for proxy traffic
+const PPV2_PROXY: u8 = 0x21;
+
+// Protocol families
+const LOCAL_PROTOCOL: u8 = 0x00;
+const TCP_IPV4: u8 = 0x11;
+const TCP_IPV6: u8 = 0x21;
+const TCP_UNIX: u8 = 0x31;
+const UDP_IPV4: u8 = 0x12;
+const UDP_IPV6: u8 = 0x22;
+const UDP_UNIX: u8 = 0x32;
+
+// Address sizes
+const IPV4_SIZE: u16 = 12;
+const IPV6_SIZE: u16 = 36;
+const UNIX_SIZE: u16 = 216;
+
+/// Read a proxy protocol event and unwrap the address information associated.
+///
+pub async fn read_proxy_protocol_event<T: TSocket>(
+    original_addr: &NetworkAddress,
+    stream: &mut T,
+) -> io::Result<NetworkAddress> {
+    // This is small enough that it should not be fragmented by TCP
+    let header: &mut [u8; 12] = &mut [0; 12];
+    stream.read_exact(header).await?;
+
+    // If it's not proxy protocol, let's stop
+    if header != &PPV2_SIGNATURE {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Invalid proxy protocol signature",
+        ));
+    }
+
+    // High 4 bits is version, low 4 bits is command
+    let version_and_command: &mut [u8; 1] = &mut [0; 1];
```

Comment:
note, fwiw, at this layer the underlying TcpSocket isn't buffered, so we're gonna be making like 5 separate read syscalls

Reply:
Yeah, I can change it to 2 reads instead of the 5.

Reply:
Changed to 2 calls now.  One to read the header, and another to read the address space.